libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/gs_seg_helpers.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <stdint.h>
+#include <setjmp.h>
+#include <string.h>
+
+#include <eosio/chain/webassembly/eos-vm-oc/eos-vm-oc.h>
+
+#ifdef __clang__
+   #define GS_PTR __attribute__((address_space(256)))
```

Comment:
I'm pretty sure GS segment is completely unused by Linux userland bits: TLS, SSP, all on FS. Arch actually has SSP enabled by default for gcc & clang and I haven't seen any problems; the asm I've seen in functions def uses FS.

Reply:
```c++
/* SPDX-License-Identifier: GPL-2.0 */
/*
 * GCC stack protector support.
 *
 * Stack protector works by putting predefined pattern at the start of
 * the stack frame and verifying that it hasn't been overwritten when
 * returning from the function.  The pattern is called stack canary
 * and unfortunately gcc requires it to be at a fixed offset from %gs.
 * On x86_64, the offset is 40 bytes and on x86_32 20 bytes.  x86_64
 * and x86_32 use segment registers differently and thus handles this
 * requirement differently.
 *
 * On x86_64, %gs is shared by percpu area and stack canary.  All
 * percpu symbols are zero based and %gs points to the base of percpu
 * area.  The first occupant of the percpu area is always
 * fixed_percpu_data which contains stack_canary at offset 40.  Userland
 * %gs is always saved and restored on kernel entry and exit using
 * swapgs, so stack protector doesn't add any complexity there.
 *
 * On x86_32, it's slightly more complicated.  As in x86_64, %gs is
 * used for userland TLS.  Unfortunately, some processors are much
 * slower at loading segment registers with different value when
 * entering and leaving the kernel, so the kernel uses %fs for percpu
 * area and manages %gs lazily so that %gs is switched only when
 * necessary, usually during task switch.
 *
 * As gcc requires the stack canary at %gs:20, %gs can't be managed
 * lazily if stack protector is enabled, so the kernel saves and
 * restores userland %gs on kernel entry and exit.  This behavior is
 * controlled by CONFIG_X86_32_LAZY_GS and accessors are defined in
 * system.h to hide the details.
 */
```
Also, Clang `ShadowCallStack` uses the gs register for this same kind of purpose.

Reply:
Your quoted blob is from kernel sources. Kernel has its own GS value; it even says so in that blob
> Userland %gs is always saved and restored on kernel entry and exit using swapgs

If you compile with `-fstack-protector-strong` you'll see code like
```
00000000000011f0 <fooz>:
fooz():
    11f0:       41 54                   push   %r12
    11f2:       48 89 fe                mov    %rdi,%rsi
    11f5:       48 83 ec 50             sub    $0x50,%rsp
    11f9:       64 48 8b 04 25 28 00    mov    %fs:0x28,%rax
...
    122c:       48 8b 44 24 48          mov    0x48(%rsp),%rax
    1231:       64 48 33 04 25 28 00    xor    %fs:0x28,%rax
    1238:       00 00 
    123a:       75 07                   jne    1243 <fooz+0x53>
    123c:       48 83 c4 50             add    $0x50,%rsp
    1240:       41 5c                   pop    %r12
    1242:       c3                      retq   
    1243:       e8 08 fe ff ff          callq  1050 <__stack_chk_fail@plt>
```
it's def using FS.

Having trouble finding applicably of ShadowCallStack, docs say it's ARM8 only
http://releases.llvm.org/9.0.0/tools/clang/docs/ShadowCallStack.html